src/Analyzers/CSharp/Analyzers/RemoveUnreachableCode/RemoveUnreachableCodeHelpers.cs

```diff
@@ -55,37 +69,87 @@ public static ImmutableArray<ImmutableArray<StatementSyntax>> GetSubsequentUnrea
-                // on that label itself  and we can use that diagnostic to handle it and any 
-                // subsequent sections.
-                break;
-            }
-
             if (currentStatement.IsKind(SyntaxKind.LocalFunctionStatement))
             {
                 // In the case of local functions, it is legal for a local function to be declared
                 // in code that is otherwise unreachable.  It can still be called elsewhere.  If
                 // the local function itself is not called, there will be a particular diagnostic
                 // for that ("The variable XXX is declared but never used") and the user can choose
-                // if they want to remove it or not. 
+                // if they want to remove it or not.
+                ConsumeCurrentSection();
+                continue;
+            }
+
+            if (i > firstUnreachableStatementIndex)
+            {
+                if (currentStatement.IsKind(SyntaxKind.LabeledStatement))
+                {
+                    // In the case of a subsequent labeled statement, we don't want to consider it 
+                    // unreachable as there may be a 'goto' somewhere else to that label.  If the 
+                    // compiler actually thinks that label is unreachable, it will give an diagnostic 
+                    // on that label itself  and we can use that diagnostic to handle it and any 
+                    // subsequent sections.
+                    break;
+                }
+
+                currentSection.Add(currentStatement);
+                continue;
+            }
+
+            if (currentStatement.IsKind(SyntaxKind.Block))
+            {
+                ProcessBlock((BlockSyntax)currentStatement);
+                continue;
+            }
+
+            void ProcessBlock(BlockSyntax block)
+            {
+                // In the case of raw blocks that are not part of other statements, like if, for, etc.
+                // we want to report all its contained statements as distinct unreachable segments, but
+                // only the statements that come after the first unreachable statement
+                ConsumeCurrentSection();
+
+                foreach (var statement in block.Statements)
+                {
+                    if (statement.IsKind(SyntaxKind.Block))
+                    {
+                        var innerBlock = (BlockSyntax)statement;
+                        if (innerBlock.SpanStart > firstStatementSpan.Start || innerBlock.Span.Contains(firstStatementSpan))
+                        {
+                            ProcessBlock(innerBlock);
+                        }
+                        continue;
+                    }
+
+                    if (statement.SpanStart > firstStatementSpan.Start)
+                    {
+                        currentSection.Add(statement);
+                    }
+                }
+
+                ConsumeCurrentSection();
+            }
```

Comment:
honestly, i just find all of this confusing and unclear as to what it is trying to accomplish and why it is the right thing.

Reply:
As the original issue notes, we might have blocks that contain statements, which may contain statements that we intend to mark as unreachable. This logic is responsible for ensuring blocks and their inner statements are properly identified as part of the control flow that has been deemed as unreachable at the point of the first unreachable statement onwards. 

We have to traverse the inner blocks' statements for the first unreachable statement. Subsequent blocks can be safely marked as unreachable thereafter, including their nested functions that are local to the unreachable blocks' scopes.